Adding System.Collections.NonGenerics ArrayList performance tests. (#18765)

KostaVlev opened the pull request:
Adding ArrayList performance tests (#18765)

stephentoub commented on src/System.Runtime.Extensions/tests/Performance/Perf.ArrayList.cs:
```diff
@@ -0,0 +1,454 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
The ArrayList tests are all in System.Collections.NonGenerics\tests.  These should probably live there, too, until such time as they're all moved.

KostaVlev replied:
I am moving it .....

stephentoub commented on src/System.Runtime.Extensions/tests/Performance/Perf.ArrayList.cs:
```diff
@@ -0,0 +1,454 @@
+                {
+                    for (int i = 0; i < Benchmark.InnerIterationCount; i++)
+                    {
+                        temp = elements.Count;
```
If it can't already, I think it's possible at some point in the future the JIT will be able to devirtualize the Count call, at which point the JIT might eliminate this whole thing.  I'd suggest changing `temp` to instead be a static volatile field.

KostaVlev replied:
Hope I fixed this correctly...

stephentoub commented on src/System.Runtime.Extensions/tests/Performance/Perf.ArrayList.cs:
```diff
@@ -0,0 +1,454 @@
+                {
+                    for (int i = 0; i < Benchmark.InnerIterationCount; i++)
+                    {
+                        var newElements = elements.GetRange(startIndex, size);
```
Nit: the `var newElements` isn't necessary

KostaVlev replied:
Fixed.

danmosemsft commented:
Thanks @KostaVlev , can you verify how long each test takes - I believe it's supposed to be 100ms-1000ms each test. (Not each iteration)

KostaVlev commented on src/System.Runtime.Extensions/tests/Performance/Perf.ArrayList.cs:
```diff
@@ -0,0 +1,454 @@
+                var elements = new ArrayList(arrayList);
+                using (iteration.StartMeasurement())
+                {
+                    for (int element = 0; element < Benchmark.InnerIterationCount; element++)
```
No problem to get this done ....

KostaVlev commented on src/System.Runtime.Extensions/tests/Performance/Perf.ArrayList.cs:
```diff
@@ -0,0 +1,454 @@
+            ArrayList arrayList = CreateArrayListOfInts(100000);            
+            foreach (var iteration in Benchmark.Iterations)
+            {
+                var elements = new ArrayList(arrayList);
```
It should be the late hour when I wrote it :).... fixed.

KostaVlev commented:
@danmosemsft They are all in between 100-1000ms.

danmosemsft commented on src/System.Collections.NonGeneric/tests/Performance/Perf.ArrayList.cs:
```diff
@@ -0,0 +1,476 @@
+        [Benchmark(InnerIterationCount = 2000000)]
+        public void Add()
+        {
+            int innerIteractionCount = (int)Benchmark.InnerIterationCount;
```
nit, some places 'Iteraction' instead of "Interaction"

KostaVlev replied:
I actually wanted to say "Iteration". :))

stephentoub commented on src/System.Collections.NonGeneric/tests/Performance/Perf.ArrayList.cs:
```diff
@@ -0,0 +1,476 @@
+{
+    public class Perf_ArrayList
+    {
+        private static volatile int _temp;
```
Nit: s_temp

stephentoub commented on src/System.Collections.NonGeneric/tests/Performance/Perf.ArrayList.cs:
```diff
@@ -0,0 +1,476 @@
+                {
+                    for (int element = 0; element < innerIterationCount; element++)
+                    {
+                        elements.Add(element);
```
Since Add takes an object rather than being strongly typed, this is going to box each int, which means we'll also be measuring the costs associated with 2,000,000 allocations.  It might be better to just allocate a single object up front, and then add that same object each time rather than adding element.

KostaVlev replied:
This is fixed now... I wish I thought about it earlier :)

stephentoub commented on src/System.Collections.NonGeneric/tests/Performance/Perf.ArrayList.cs:
```diff
@@ -0,0 +1,476 @@
+                for (int index = 0; index < elementsCollectionSize; index++)
+                {
+                    elementsCollection[index] = new ArrayList(elements);
+                }                    
```
I wonder if it's worthwhile doing similar prep for the AddRange case earlier.  Might not be worthwhile given that in that case the prep is just the allocation/ctor for the temporary ArrayList.

KostaVlev replied:
That's not bad idea. I am fixing the AddRange test.

stephentoub commented on src/System.Collections.NonGeneric/tests/Performance/Perf.ArrayList.cs:
```diff
@@ -0,0 +1,476 @@
+            ArrayList elements = new ArrayList(size);
+            for (int index = 0; index < size; index++)
+            {
+                elements.Add(random.Next());
```
What's the purpose of the randomness here?  Are the actual values used somewhere?

KostaVlev replied:
amm... yes I am fixed this too....

stephentoub commented on src/System.Collections.NonGeneric/tests/Performance/Perf.ArrayList.cs:
```diff
@@ -0,0 +1,476 @@
+        {
+            int innerIterationCount = (int)Benchmark.InnerIterationCount;
+            int size = (int)Benchmark.InnerIterationCount;           
+            var random = new Random();
```
The randomness here can affect the timing, since the values used can affect how many comparison operations are needed in the sort and thus the running time of the sort.  It'd be best to use a fixed seed with the Random ctor, so that all runs employ the same values and are deterministic.

danmosemsft replied:
Is there value in having a benchmark for a "pathological" input, similar to what you fixed in Linq recently? Eg., sorted input. I'm guessing probably not, since it would only matter if someone changed the sort algorithm, and they're not going to do that without measuring..

stephentoub replied:
ArrayList.Sort just uses Array.Sort.  We should have a variety of tests for Array.Sort, but I think this is sufficient for ArrayList.Sort.

KostaVlev replied:
This is fixed too...

KostaVlev commented:
I think I addressed all the reviews. I am attaching and tests results below..  

![perf-tests](https://user-images.githubusercontent.com/10883826/31420895-3b2471e4-ae12-11e7-817c-25bb10d6f73f.png)
